File: src/com/timepath/hl2/io/studiomodel/StudioModel.java
package com.timepath.hl2.io.studiomodel;

import java.io.*;
import java.nio.*;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class StudioModel {

    private static final Logger LOG = Logger.getLogger(StudioModel.class.getName());

    static final int MAX_NUM_BONES_PER_VERT = 3;

    static final int MAX_NUM_LODS = 8;

    public final MDL mdl;

    public final VTX vtx;

    public final VVD vvd;

    private final ByteBuffer indexBuffer;

    public StudioModel(InputStream mdlStream, InputStream vvdStream, InputStream vtxStream) throws IOException {
        mdl = MDL.load(mdlStream);
        vvd = VVD.load(vvdStream);
        vtx = VTX.load(vtxStream);
        int lod = 0;
        setRootLOD(lod);
        indexBuffer = buildIndices(lod);
    }

    public IntBuffer getIndices() {
        return indexBuffer.asIntBuffer();
    }

    public FloatBuffer getNormals() {
        return vvd.normals.asFloatBuffer();
    }

    public FloatBuffer getTangents() {
        return vvd.tangents.asFloatBuffer();
    }

    public FloatBuffer getTextureCoordinates() {
        return vvd.uv;
    }

    public FloatBuffer getVertices() {
        return vvd.verts.asFloatBuffer();
    }

    private ByteBuffer buildIndices(int lodId) {
        ByteArrayOutputStream indices = new ByteArrayOutputStream();
        int indexOffset = 0;
        int vertTableIndex;

        for(int i = 0; i < vtx.bodyParts.size(); i++) {
            VTX.BodyPart bodyPart = vtx.bodyParts.get(i);
            MDL.MStudioBodyParts mdlBodyPart = mdl.mdlBodyParts.get(i);

            if(bodyPart.models.isEmpty()) {
                continue;
            }
            VTX.Model model = bodyPart.models.get(0);
            MDL.MStudioModel mdlModel = mdlBodyPart.models.get(0);

            VTX.ModelLOD lod = model.lods.get(lodId);

            for(int j = 0; j < lod.meshes.size(); j++) {
                VTX.Mesh mesh = lod.meshes.get(j);
                MDL.MStudioMesh mdlMesh = mdlModel.meshes.get(j);

                for(VTX.StripGroup stripGroup : mesh.stripGroups) {
                    List<VTX.Vertex> vertTable = stripGroup.verts;
                    stripGroup.indexOffset = indexOffset++;
                    ShortBuffer sb = stripGroup.indexBuffer.order(ByteOrder.LITTLE_ENDIAN).asShortBuffer();
                    for(int l = 0; l < stripGroup.numIndices; l++) {
                        vertTableIndex = sb.get();
                        int index = vertTable.get(vertTableIndex).origMeshVertID
                                        + mdlModel.vertexoffset + mdlMesh.vertexoffset;
                        short s = (short) index;
                        try {
                            indices.write(new byte[] {
                                (byte) (s & 0xFF),
                                (byte) ((s & 0xFF00) >> 8),
                                (byte) ((s & 0xFF0000) >> 16),
                                (byte) ((s & 0xFF000000) >> 24),});
                        } catch(IOException ex) {
                            LOG.log(Level.SEVERE, null, ex);
                            return null;
                        }
                    }
                }
            }
        }
        byte[] bytes = indices.toByteArray();
        ByteBuffer buf = ByteBuffer.allocateDirect(bytes.length);
        buf.put(bytes).flip();
        return buf;
    }

    private void setRootLOD(int rootLOD) {
        MDL.StudioHeader header = mdl.header;
        List<MDL.MStudioBodyParts> bodyParts = mdl.mdlBodyParts;

        if(header.numAllowedRootLODs > 0 && rootLOD >= header.numAllowedRootLODs) {
            rootLOD = header.numAllowedRootLODs - 1;
        }

        int vertexoffset = 0;

        for(MDL.MStudioBodyParts bodyPart : bodyParts) {
            for(MDL.MStudioModel model : bodyPart.models) {
                int totalMeshVertices = 0;
                for(int meshId = 0; meshId < model.meshes.size(); ++meshId) {
                    MDL.MStudioMesh mesh = model.meshes.get(meshId);

                    mesh.numvertices = mesh.vertexdata.numLODVertexes[rootLOD];
                    mesh.vertexoffset = totalMeshVertices;
                    totalMeshVertices += mesh.numvertices;
                }

                model.numvertices = totalMeshVertices;
                model.vertexoffset = vertexoffset;
                vertexoffset += totalMeshVertices;
            }
        }
    }

}

File: src/main/kotlin/com/timepath/hl2/io/studiomodel/StudioModel.kt
package com.timepath.hl2.io.studiomodel


import com.timepath.Logger
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.IntBuffer
import java.util.logging.Level

public class StudioModel @throws(IOException::class) constructor
(mdlStream: InputStream, vvdStream: InputStream, vtxStream: InputStream) {
    public val mdl: MDL?
    private val vtx: VTX?
    private val vvd: VVD?
    private val indexBuffer: ByteBuffer?

    init {
        mdl = MDL.load(mdlStream)
        vvd = VVD.load(vvdStream)
        vtx = VTX.load(vtxStream)
        val lod = 0
        setRootLOD(lod)
        indexBuffer = buildIndices(lod)
    }

    private fun buildIndices(lodId: Int): ByteBuffer? {
        val indices = ByteArrayOutputStream()
        var indexOffset = 0
        for (i in vtx!!.bodyParts.indices) {
            val bodyPart = vtx.bodyParts[i]
            val mdlBodyPart = mdl!!.mdlBodyParts[i]
            if (bodyPart.models.isEmpty()) {
                continue
            }
            val model = bodyPart.models[0]
            val mdlModel = mdlBodyPart.models!![0]
            val lod = model.lods[lodId]
            for (j in lod.meshes.indices) {
                val mesh = lod.meshes[j]
                val mdlMesh = mdlModel.meshes!![j]
                for (stripGroup in mesh.stripGroups) {
                    val vertTable = stripGroup.verts
                    stripGroup.indexOffset = indexOffset++
                    val sb = stripGroup.indexBuffer.order(ByteOrder.LITTLE_ENDIAN).asShortBuffer()
                    for (l in 0..stripGroup.numIndices - 1) {
                        val vertTableIndex = sb.get().toInt()
                        val index = vertTable[vertTableIndex].origMeshVertID.toInt() + mdlModel.vertexoffset + mdlMesh.vertexoffset
                        val s = index.toShort()
                        try {
                            indices.write(byteArrayOf((s.toInt() and 0xFF).toByte(), ((s.toInt() and 0xFF00) shr 8).toByte(), ((s.toInt() and 0xFF0000) shr 16).toByte(), ((s.toInt() and -16777216) shr 24).toByte()))
                        } catch (ex: IOException) {
                            LOG.log(Level.SEVERE, { null }, ex)
                            return null
                        }

                    }
                }
            }
        }
        val bytes = indices.toByteArray()
        val buf = ByteBuffer.allocateDirect(bytes.size())
        buf.put(bytes).flip()
        return buf
    }

    private fun setRootLOD(rootLOD: Int) {
        @suppress("NAME_SHADOWING")
        var rootLOD = rootLOD
        mdl!!
        val header = mdl.header
        val bodyParts = mdl.mdlBodyParts
        if ((header.numAllowedRootLODs > 0) && (rootLOD >= header.numAllowedRootLODs)) {
            rootLOD = header.numAllowedRootLODs.toInt() - 1
        }
        var vertexoffset = 0
        for (bodyPart in bodyParts) {
            for (model in bodyPart.models!!) {
                var totalMeshVertices = 0
                for (meshId in model.meshes!!.indices) {
                    val mesh = model.meshes!![meshId]
                    mesh.numvertices = mesh.vertexdata.numLODVertexes[rootLOD]
                    mesh.vertexoffset = totalMeshVertices
                    totalMeshVertices += mesh.numvertices
                }
                model.numvertices = totalMeshVertices
                model.vertexoffset = vertexoffset
                vertexoffset += totalMeshVertices
            }
        }
    }

    public val indices: IntBuffer
        get() = indexBuffer!!.asIntBuffer()

    public val normals: FloatBuffer
        get() = vvd!!.normalBuffer.asFloatBuffer()

    public val tangents: FloatBuffer
        get() = vvd!!.tangentBuffer.asFloatBuffer()

    public val textureCoordinates: FloatBuffer
        get() = vvd!!.uvBuffer

    public val vertices: FloatBuffer
        get() = vvd!!.vertexBuffer.asFloatBuffer()

    companion object {

        val MAX_NUM_BONES_PER_VERT = 3
        val MAX_NUM_LODS = 8
        private val LOG = Logger()
    }
}
